### DISCUSSION TITLE
how to change mountpoint for referenced photos

### DISCUSSION THREAD

**Discussion by jgc234** [RhetTbull/osxphotos#1664] [Q&A]
I'd like to change the mounted file sharing protocol of 300,000 referenced photos from AFP to SMB. I've gone down a rabbit hole and believe I've found the protocol and mountpoint hidden in some awful binary mess in ZFILESYSTEMBOOKMARK.ZBOOKMARKDATA for each asset. I'm trying to hack up update_photos_bookmarks.py to adjust this, but I don't really understand the world of apple of CFURL thingos just yet.  

Any thoughts on how to do this?  Is there another, simpler, way to alter the referenced URL prefix for all photos?

**Comment by RhetTbull** [comment]
This is a very difficult problem because Apple Photos is basically broken when it comes to referenced photos. Because of this I advise people to never use referenced photos. In the current implementation, Photos uses [security scoped bookmarks](https://developapa.com/security-scoped-bookmark/) for referenced files. This is same as any sandboxed macOS app would need to do to access a file outside of its sandbox. However, these bookmarks are signed so they can only be used by the app that has them and they cannot be changed.  You could replace `ZBOOKMARKDATA` but the signature of the bookmark wouldn't match and Photos would refuse to use it (I've tried). 

> I'm trying to hack up update_photos_bookmarks.py to adjust this

I think you're referring to [RepairPhotosBookmarks](https://github.com/RhetTbull/RepairPhotosBookmarks) which is a hack of a script that attempts to fix this. I have plans to eventually roll this capability into OSXPhotos. The script hasn't been updated in some time and may not work on your version of macOS. What the script does is read the bookmark data to get the path then creates a new empty library and imports the photo into the library. This causes Photos to create the bookmark which is then copied from the temporary library into the old library. This is a hack and may or may not work in your case. Some users have had success but others have run into problems and I don't have time to help diagnose this.

> Is there another, simpler, way to alter the referenced URL prefix for all photos?

If there was I'd have implemented it. I'm not aware of any other way to do fix the bookmarks.

Here's an alternative workflow that would probably work with some caveats.

Use osxphotos to export the photos to the new mount point. Export the AAE files (which include information to reconstruct edits). If you use  a specific directory structure you'd need to use `--directory` with a template to reconstruct this. 

`osxphotos export /path/to/mount/point --export-aae`

Move all the existing photos to an album so you can remove them later.

Import all the photos using `osxphotos import`. Use the `--exportdb` option which will read metadata from the export database created by OSXPhotos in step 1 to recreate all metadata (albums, keywords, titles, captions) with exception of faces/persons (Photos limitation). 

`osxphotos import /path/to/mount/point --walk --exportdb /path/to/mount/point `

This re-imports everything into library which will create correct bookmarks then re-applies all the metadata. 

Go into the album you created above and delete all those photos with the old bookmark.

Here's the big caveat that might make this unacceptable to you: Even if you have "copy items to the Photos library" disabled in `Photos > Settings` (e.g. a referenced library), any edited images that are imported with the edited preview and the AAE file to preserve the non-destructive edits, will be copied to the library. This is a behavior that Photos implements and I've tried unsuccessfully to defeat it. Thus while you'll preserve all edits, any images that have been edited will no longer be referenced.

**Reply by RhetTbull** [reply]
To get the original referenced path for creating the export you could do something like this:

Let's assume your old mount point was `/Volumes/OldPhotos/` which is 20 characters long:

`--directory "{photo.isreference?{photo.path|chomp(20)},}"`

This cuts the first 20 characters off the photo path for referenced photos leaving just the original sub-path and uses that as the new path in the export folder. If you have any non-referenced photos you'd need to provide a default value to use in that case. For example, this would organize those by `YYYY/MM/DD`:

`--directory "{photo.isreference?{photo.path|chomp(20)},{created.year}/{created.mm}/{created.dd}}"`

**Reply by jgc234** [reply]
Thank you for the detailed replies.  That's a tonne of useful information to keep me entertained for a while.

Ultimately I may need to heed your advice and change my strategy away from referenced photos.  The original idea was to use centralised storage (already reliable, high performance, snapshots, offsite replication, etc), rather than maintaining yet another unprotected large local storage pool and managing separate backups.  The durability of the storage is offset by Photos periodic crashing, potentially from the yo-yo of mounts during power save and wake up.  Nice idea, until it got large and unwieldy.

**Reply by cfc62** [reply]
Photos is... not great with centralized storage. The Apple ecosystem for Photos is not just the Photos app, but a ton of other daemons and services and helper programs that wants/needs access to your photos at any time. 

photoanalysisd, mediaanalysisd, biome, even spotlight really dig into your library, and the depth and breadth of your Mac looking at your photos is just going to get deeper and more involved over time. 

Apple semi-quietly removed a lot of analysis on local drives, such as SSDs over Thunderbolt - now for the full analysis package your library (and photos) needs to be completely contained onto your internal drive. 

A logical question is "how much does this analysis benefit me?" and so far, not a whole lot, but I suspect that's about to change. 

I don't like this, I think it's an artificial limitation but there you go. Of note, Lightroom Classic also does not work well on NASs, so this isn't just an Apple thing, these apps are meant for single-user at a time usage. 

I've never had any material problems with Photos unless I was trying something way out of the ordinary. It's quite stable sitting on my MBP and backup is just using Finder to copy the library bundle to an SSD, then moving that SSD to 2 other Macs (one after the other) and copying the library onto those machines. It's not 1-2-3 (I don't have 2 locations) but I'm comfortable with that.

**Reply by cfc62** [reply]
Sorry meant to include this parting note:

In the end I decided not to fight what Apple wants you to do with Photos but rather to go with the flow, hence my working library fits entirely onto my MBP and edits are done there. Apple fires up its processes and analyzes my content on a pretty consistent basis. I used to fight this too but now I just let it do its thing. 

I've got well over 100,000 assets and it's over 1.5T now for library size. You clearly have a bigger library, probably one of the biggest ones I've heard of. Many people get nervous around 10,000 assets and I chuckle at that.

**Reply by RhetTbull** [reply]
> Apple semi-quietly removed a lot of analysis on local drives, such as SSDs over Thunderbolt - now for the full analysis package your library (and photos) needs to be completely contained onto your internal drive.

I'm sure there's a great technical reason for this and not
That this is a ploy to drive you to pay for more overpriced internal storage...

**Reply by cfc62** [reply]
One "reason" I've heard from Apple is for "privacy" of derived data. 

(I'd have to hunt around for the citation, sorry)

Uhhh, OK, but my external thunderbolt SSD is encrypted...

**Reply by cfc62** [reply]
Oh also I tend to believe Photos is very much relying on APFS, so moving to SMB may be problematic just on that basis alone.
